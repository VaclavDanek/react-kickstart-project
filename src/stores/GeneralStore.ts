import moment from 'moment'
import { makeAutoObservable } from 'mobx'
import { ErrorTypesEnum } from '../types/errorTypes'

// types
import type { AxiosError, CustomErrorEvent, CustomError } from '../types/errorTypes'
import type { Alert } from '../types'

class GeneralStore {
  alerts: Alert[] = []
  errors: CustomError[] = []
  fetching: number = 0

  constructor() {
    makeAutoObservable(this)
  }

  addAlert(alert: Alert): void {
    this.alerts.push(alert)
  }

  removeAlert(index: number): void {
    this.alerts.splice(index, 1)
  }

  addError(error: CustomError): void {
    this.errors.push(error)
  }

  removeError(index: number): void {
    this.errors.splice(index, 1)
  }

  setFetching(value: number): void {
    if (value >= 0) {
      this.fetching = value
    }
  }

  startFetching(): void {
    ++this.fetching
  }

  stopFetching(): void {
    if (this.fetching > 0) {
      --this.fetching
    }
  }

  onActionFailure(errorEvent: ErrorEvent | CustomErrorEvent, errorType: ErrorTypesEnum = ErrorTypesEnum.WARNING): void {
    this.addError({
      error: { ...errorEvent },
      type: errorType,
      timestamp: moment().format('YYYY-MM-DD HH:mm:ss'),
    });
    if (process.env.NODE_ENV === 'development') {
      console.error(errorEvent.message, errorType);
    }
  }

  onRequestFailure(error: AxiosError, errorType: ErrorTypesEnum = ErrorTypesEnum.WARNING): void {
    const { message, response, stack } = error;
    if (message === 'Network Error' || response?.status === 401) { // ignore connection lost and 401 cause it is probably just unsuccessful authentication...
      return
    }

    this.addError({
      error: { ...error, message, response, stack },
      type: errorType,
      timestamp: moment().format('YYYY-MM-DD HH:mm:ss'),
    })
    if (process.env.NODE_ENV === 'development') {
      console.error(error.message, errorType)
    }
  }
}

const generalStore = new GeneralStore()
export default generalStore
