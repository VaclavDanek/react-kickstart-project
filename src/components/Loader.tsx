import { default as Loader } from 'react-loaders'
import { observer } from 'mobx-react'

// stores
import GeneralStore from '../stores/GeneralStore'

// types
import type { JSX } from 'react'
import type { LoaderType } from 'react-loaders'

// styles
import 'loaders.css/src/animations/ball-pulse-sync.scss'
// import 'loaders.css/src/animations/ball-pulse.scss'
// import 'loaders.css/src/animations/ball-clip-rotate.scss'
// import 'loaders.css/src/animations/ball-scale-multiple.scss'
// import 'loaders.css/src/animations/line-scale.scss'

interface LoaderProps {
  className?: string;
  innerClassName?: string;
  active?: boolean;
  type?: LoaderType;
  color?: string;
}

export default observer((props: Readonly<LoaderProps>): JSX.Element | null => {
  const { 
    className = 'vh-center', 
    type = 'ball-pulse-sync', 
    color = '#0d6efd', 
    active = GeneralStore.fetching > 0, 
    innerClassName, 
  } = props

  return !active ? null : (
    <div className='bg-white bg-opacity-80 w-full h-full top-0 absolute'>
      <Loader
        {...{
          className,
          innerClassName,
          active,
          type,
          color,
        }}
      />
    </div>
  )
})
