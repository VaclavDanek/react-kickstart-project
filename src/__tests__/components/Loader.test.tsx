import renderer from 'react-test-renderer'

// components
import { Loader } from '../../components'

// stores
import GeneralStore from '../../stores/GeneralStore'

describe('Loader', () => {
  it('should render null when is not fetching', () => {
    const tree = renderer.create(<Loader active={GeneralStore.fetching > 0} />).toJSON();
    expect(tree).toBeNull();
  });

  it('should render Loader while fetching', () => {
    GeneralStore.startFetching();
    const tree = renderer.create(<Loader active={GeneralStore.fetching > 0} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});