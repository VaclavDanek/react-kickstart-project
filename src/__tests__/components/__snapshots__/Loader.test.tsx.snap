// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoaderWrapper should render the loader with custom props 1`] = `
<div
  className="bg-white bg-opacity-80 w-full h-full top-0 absolute"
>
  <div
    className="loader loader-undefined loader-active custom-wrapper"
  >
    <div
      className="loader-inner pacman custom-inner"
    >
      <div
        style={
          {
            "backgroundColor": "red",
          }
        }
      />
      <div
        style={
          {
            "backgroundColor": "red",
          }
        }
      />
      <div
        style={
          {
            "backgroundColor": "red",
          }
        }
      />
      <div
        style={
          {
            "backgroundColor": "red",
          }
        }
      />
      <div
        style={
          {
            "backgroundColor": "red",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`LoaderWrapper should render the loader with default props when active is true 1`] = `
<div
  className="bg-white bg-opacity-80 w-full h-full top-0 absolute"
>
  <div
    className="loader loader-undefined loader-active vh-center"
  >
    <div
      className="loader-inner ball-pulse-sync"
    >
      <div
        style={
          {
            "backgroundColor": "#0d6efd",
          }
        }
      />
      <div
        style={
          {
            "backgroundColor": "#0d6efd",
          }
        }
      />
      <div
        style={
          {
            "backgroundColor": "#0d6efd",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`LoaderWrapper should render while fetching 1`] = `
<div
  className="bg-white bg-opacity-80 w-full h-full top-0 absolute"
>
  <div
    className="loader loader-undefined loader-active vh-center"
  >
    <div
      className="loader-inner ball-pulse-sync"
    >
      <div
        style={
          {
            "backgroundColor": "#0d6efd",
          }
        }
      />
      <div
        style={
          {
            "backgroundColor": "#0d6efd",
          }
        }
      />
      <div
        style={
          {
            "backgroundColor": "#0d6efd",
          }
        }
      />
    </div>
  </div>
</div>
`;
